File: 13/python/main.py
from collections import defaultdict

def fold_up(input_points, at):
    result = []
    for (x,y) in input_points:
        if y < at:
            result.append((x,y))
        else:
            result.append((x, at - (y - at)))
    return result

def fold_left(input_points, at):
    result = []
    for (x,y) in input_points:
        if x < at:
            result.append((x,y))
        else:
            result.append((at - (x - at), y))
    return list(set(result))

def solution(lines):
   print(lines)
   [input_points, instructions] = (lines).split('\n\n')
   input_points = [x.split(",") for x in input_points.split("\n")]
   input_points = [(int(x), int(y)) for [x,y] in input_points]

   instructions = [x.split("=") for x in instructions.split("\n")]
   instructions = [(x[-1], int(y)) for x,y in instructions]
   print(instructions)

   for dir, at in instructions:
       if dir == 'x':
           input_points = fold_left(input_points, at)
        
       if dir == 'y':
           input_points = fold_up(input_points, at)
   max_x = max(x for (x, _) in input_points)
   max_y = max(x for (_, x) in input_points)
   for y in range(max_y+1):
       for x in range(max_x+1):
           if (x,y) in input_points:
               print("#", end="")
           else:
                print(".", end="")
       print("\n", end="")

   
#    grid = [[False for x in range(2000)] for y in range(2000)]
#    for x,y in input_points:
#        grid[x][y] = True
#    print(grid)

with open('input.txt') as f:
    lines = f.read()
    solution(lines)

File: 13/haskell/main.hs
import Data.Either
import Data.List
import System.IO
import Text.ParserCombinators.Parsec

type Point = (Int, Int)

data Direction = Horizontal | Vertical deriving (Show, Eq)

data FoldInstruction = FoldInstruction Direction Int deriving (Show)

data Day13Input = Day13Input [Point] [FoldInstruction] deriving (Show)

day13 :: CharParser () Day13Input
day13 =
  Day13Input
    <$> endBy point (char '\n')
    <*> (char '\n' >> sepBy instruction (char '\n'))

instruction :: CharParser () FoldInstruction
instruction =
  FoldInstruction
    <$> (string "fold along " >> direction)
    <*> (char '=' >> number)

direction :: CharParser () Direction
direction =
  Horizontal <$ char 'x'
    <|> Vertical <$ char 'y'

point :: CharParser () Point
point = (,) <$> number <*> (char ',' >> number)

number :: CharParser () Int
number = read <$> many1 digit

applyInstruction :: FoldInstruction -> [Point] -> [Point]
applyInstruction (FoldInstruction Horizontal at) =
  nub . map (\(x, y) -> (at - abs (at - x), y))
applyInstruction (FoldInstruction Vertical at) =
  nub . map (\(x, y) -> (x, at - abs (at - y)))

solution1 :: Day13Input -> Int
solution1 (Day13Input points instructions) =
  length $ applyInstruction (head instructions) points

prettyPrintPoints :: [Point] -> String
prettyPrintPoints points =
  [ if x == maxX + 1 then '\n' else if (x, y) `elem` points then '#' else '.'
    | y <- [0 .. maxY],
      x <- [0 .. maxX + 1]
  ]
  where
    maxX = maximum $ map fst points
    maxY = maximum $ map snd points

solution2 :: Day13Input -> String
solution2 (Day13Input points instructions) =
  prettyPrintPoints $ foldl (flip applyInstruction) points instructions

main = do
  handle <- openFile "input.txt" ReadMode
  contents <- hGetContents handle
  let parsed = fromRight (Day13Input [] []) $ parse day13 "Invalid input" contents
  print (solution1 parsed)
  putStrLn (solution2 parsed)
  hClose handle
